### PR TITLE
Graceful restart local ssh daemon

### PR DESCRIPTION
**Pull request by mustard-mh**
- [ ] It should rebase after PR https://github.com/gitpod-io/gitpod-vscode-desktop/pull/50 merged

This PR changed 

#### 1️⃣  add restart logic in daemon
- If there's no connection to local-ssh
- Get extension version from extension ipc
- Compare with daemon running extension version
- If it's greater than current one, ask extension ipc to restart daemon later (trigger `onDidDaemonRestartRequired` event)
- Exit with proper exit code

#### 2️⃣  where we start daemon
 from
- Extension active
- Extension ipc service active and every time we are trying to ping local-ssh ipc

To 
- Extension active
- Extension ipc service received `onDidDaemonRestartRequired` event
- Trying to start a remote connection

## Related Issue(s)

Fixes https://linear.app/gitpod/issue/IDE-54

### CODE REVIEW

**Comment by mustard-mh** [comment]
@jeanp413 I updated PR description, not test it yet, will continue work on it tomorrow, but you can review it first if you have time since it's almost complete 🙏

**Comment by jeanp413** [comment]
Let's discuss a little bit more:
- what's the delay of starting the daemon on demand?
- right now I only see two cases:
  - on activate **if restoring a remote window** this means user closes vscode completely so daemon is not running
  - on uriHandler when trying to connect to a new workspace, this will be the first remote window, we "save" here time if user has already open vscode but hown much? other case is vscode was completely closed and is started by the uri handler and we need to start daemon at the same time

at any other time daemon don't need to be running I think, I'm I missing some other case

File: src/local-ssh/ipc/extension.ts [hunk 1]
```diff
@@ -185,22 +214,26 @@ export class ExtensionServiceServer extends Disposable {
...
-            this.backoffActive();
+            this.setExtensionActivedForDaemon();
         }).catch(e => {
             this.logService.error(e, `extension ipc service server failed to listen with id: ${this.id}`);
         });
     }
 
-    private async backoffActive() {
+    private async setExtensionActivedForDaemon() {
+        while (true) {
```
Comment (akosyakov) [inline src/local-ssh/ipc/extension.ts:224]: should not we cancel such things that our extension don't keep doing anything if deactivated

File: src/local-ssh/server.ts [hunk 2]
```diff
@@ -59,6 +67,28 @@ export class LocalSSHGatewayServer {
 		return session;
 	}
 
+	private async tryReloadDaemon() {
+		if (this.clientCount > 0) {
```
Comment (akosyakov) [inline src/local-ssh/server.ts:71]: Does it mean we would run several daemons in parallel? i.e. if some window already upgraded and assumes a never version of daemon, but others are not yet?

Comment (mustard-mh) [reply src/local-ssh/server.ts:71]: It's connection count (need to rename it)

**Pull request closed without merging**